src/Products/ZMySQLDA/db.py

```diff
@@ -640,15 +646,25 @@ def _begin(self, *ignored):
         Also called from _register() upon first query.
         """
         try:
+            try:
+                self.db.ping()
+            except OperationalError as exc:
+                # Starting with mysqlclient version 2.2.1 the connection object
+                # ``ping`` method behavior changed. It may raise an exception.
+                # If the connection is broken attempt to reconnect first.
+                if exc.args[0] in hosed_connection:
+                    self._forceReconnection(reason=exc.args[0])
```

Comment:
We may want to raise a `transaction.interfaces.TransientError` here to get the transaction retried.
Such an exception might be more important at other `_forceReconnection` places. There the "hosed connection" may already have performed some changes which get lost by the reconnection. Here, we likely are before the first command against the connection in this transaction; thus, simply reconnecting might be okay.

Perhaps we should put a retrying raise into `_forceReconnection` itself when the reconnection is called for due to a "hosed connection".

Reply:
My goal was to change the behavior as little as possible. I am working off the assumption that when this `_begin` method is called there are no pending transactions on the relational database side, so reconnecting won't break those. Other Zope transaction changes, like ZODB changes, will not be affected by just silently re-opening the relational database connection. I am trying to do the least amount of interference on the currently running transaction. Do you have reason to believe that this is an unsafe way to deal with the ongoing transaction?